File: src/agent/agent-sdk/nodejs/src/httpClient.ts
/*
 * BK-CI Agent SDK - HTTP 客户端与结果封装
 *
 * 对应 Go agent 的 src/pkg/util/httputil。使用 Node 内置 http/https，零第三方依赖。
 * 统一处理鉴权头注入、请求超时、以及 DevopsResult / AgentResult 的解析。
 */

import * as http from 'http';
import * as https from 'https';
import { URL } from 'url';
import { AgentStatus } from './config';

/** 后台统一返回结构（对应 Go DevopsResult）。status===0 表示成功。 */
export interface DevopsResult<T = unknown> {
  data: T;
  status: number;
  message: string;
}

/** ask 接口返回结构（对应 Go AgentResult），额外带 agentStatus。 */
export interface AgentResult<T = unknown> extends DevopsResult<T> {
  agentStatus: string;
}

export function isOk(r: DevopsResult): boolean {
  return r.status === 0;
}

export function isNotOk(r: DevopsResult): boolean {
  return r.status !== 0;
}

export function isAgentDelete(r: AgentResult): boolean {
  return !!r.agentStatus && r.agentStatus === AgentStatus.Delete;
}

export interface HttpClientOptions {
  /** 请求超时（毫秒） */
  timeoutMs?: number;
}

export interface RequestOptions {
  method: 'GET' | 'POST' | 'PUT' | 'DELETE';
  url: string;
  headers?: Record<string, string>;
  /** 请求体对象，将被 JSON.stringify */
  body?: unknown;
  timeoutMs?: number;
}

export interface RawResponse {
  status: number;
  body: string;
}

/** 流式请求选项（用于文件下载，不缓冲响应体）。 */
export interface StreamRequestOptions {
  method: 'GET' | 'POST' | 'PUT' | 'DELETE';
  url: string;
  headers?: Record<string, string>;
  timeoutMs?: number;
}

/** 流式响应：暴露状态码、响应头以及可 pipe 的响应流（IncomingMessage）。 */
export interface StreamResponse {
  status: number;
  headers: http.IncomingHttpHeaders;
  stream: http.IncomingMessage;
}

/**
 * 发起一次流式 HTTP 请求，不缓冲响应体，返回状态码 + 响应头 + 可 pipe 的流。
 * 供文件下载（download.ts）使用。调用方负责消费/销毁 stream。
 */
export function requestStream(opts: StreamRequestOptions): Promise<StreamResponse> {
  return new Promise((resolve, reject) => {
    let parsed: URL;
    try {
      parsed = new URL(opts.url);
    } catch (e) {
      reject(new Error(`invalid url: ${opts.url}`));
      return;
    }

    const isHttps = parsed.protocol === 'https:';
    const lib = isHttps ? https : http;

    const req = lib.request(
      {
        protocol: parsed.protocol,
        hostname: parsed.hostname,
        port: parsed.port || (isHttps ? 443 : 80),
        path: parsed.pathname + parsed.search,
        method: opts.method,
        headers: { ...(opts.headers ?? {}) },
      },
      (res) => {
        resolve({
          status: res.statusCode ?? 0,
          headers: res.headers,
          stream: res,
        });
      }
    );

    const timeout = opts.timeoutMs ?? 300_000;
    req.setTimeout(timeout, () => {
      req.destroy(new Error(`request timeout after ${timeout}ms: ${opts.url}`));
    });

    req.on('error', (err) => reject(err));
    req.end();
  });
}

/**
 * 发起一次 HTTP 请求，返回原始响应文本。基于 Node 内置模块，支持 http/https。
 */
export function request(opts: RequestOptions): Promise<RawResponse> {
  return new Promise((resolve, reject) => {
    let parsed: URL;
    try {
      parsed = new URL(opts.url);
    } catch (e) {
      reject(new Error(`invalid url: ${opts.url}`));
      return;
    }

    const isHttps = parsed.protocol === 'https:';
    const lib = isHttps ? https : http;

    const headers: Record<string, string> = { ...(opts.headers ?? {}) };
    let payload: string | undefined;
    if (opts.body !== undefined && opts.body !== null) {
      payload = JSON.stringify(opts.body);
      headers['Content-Type'] = 'application/json';
      headers['Content-Length'] = Buffer.byteLength(payload).toString();
    }

    const req = lib.request(
      {
        protocol: parsed.protocol,
        hostname: parsed.hostname,
        port: parsed.port || (isHttps ? 443 : 80),
        path: parsed.pathname + parsed.search,
        method: opts.method,
        headers,
      },
      (res) => {
        const chunks: Buffer[] = [];
        res.on('data', (c) => chunks.push(c as Buffer));
        res.on('end', () => {
          resolve({
            status: res.statusCode ?? 0,
            body: Buffer.concat(chunks).toString('utf-8'),
          });
        });
      }
    );

    const timeout = opts.timeoutMs ?? 30_000;
    req.setTimeout(timeout, () => {
      req.destroy(new Error(`request timeout after ${timeout}ms: ${opts.url}`));
    });

    req.on('error', (err) => reject(err));

    if (payload !== undefined) {
      req.write(payload);
    }
    req.end();
  });
}

/** HTTP 客户端：持有默认超时，提供解析为 DevopsResult / AgentResult 的便捷方法。 */
export class HttpClient {
  private readonly timeoutMs: number;

  constructor(opts: HttpClientOptions = {}) {
    this.timeoutMs = opts.timeoutMs ?? 30_000;
  }

  async requestRaw(opts: Omit<RequestOptions, 'timeoutMs'>): Promise<RawResponse> {
    return request({ ...opts, timeoutMs: this.timeoutMs });
  }

  async intoDevopsResult<T = unknown>(
    opts: Omit<RequestOptions, 'timeoutMs'>
  ): Promise<DevopsResult<T>> {
    const raw = await this.requestRaw(opts);
    return parseResult<DevopsResult<T>>(raw, opts.url);
  }

  async intoAgentResult<T = unknown>(
    opts: Omit<RequestOptions, 'timeoutMs'>
  ): Promise<AgentResult<T>> {
    const raw = await this.requestRaw(opts);
    return parseResult<AgentResult<T>>(raw, opts.url);
  }
}

function parseResult<R>(raw: RawResponse, url: string): R {
  try {
    return JSON.parse(raw.body) as R;
  } catch (e) {
    throw new Error(
      `parse result error, url=${url} status=${raw.status} body=${raw.body.slice(0, 500)}`
    );
  }
}

File: src/agent/agent-sdk/python/bkci_agent_sdk/http_client.py
"""Small dependency-free asynchronous HTTP client used by the SDK."""

from __future__ import annotations

import asyncio
import json
from dataclasses import asdict, dataclass, is_dataclass
from enum import Enum
from http.client import HTTPMessage
from typing import Any, BinaryIO, Callable, Generic, Mapping, TypeVar
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from .config import AgentStatus

T = TypeVar("T")


@dataclass(slots=True)
class DevopsResult(Generic[T]):
    data: T | None = None
    status: int = -1
    message: str = ""


@dataclass(slots=True)
class AgentResult(DevopsResult[T]):
    agent_status: str = ""


@dataclass(slots=True)
class RawResponse:
    status: int
    body: str


@dataclass(slots=True)
class StreamResponse:
    status: int
    headers: HTTPMessage
    stream: BinaryIO


RawTransport = Callable[..., RawResponse]


def is_ok(result: DevopsResult[Any]) -> bool:
    return result.status == 0


def is_not_ok(result: DevopsResult[Any]) -> bool:
    return result.status != 0


def is_agent_delete(result: AgentResult[Any]) -> bool:
    return bool(result.agent_status) and result.agent_status == AgentStatus.DELETE


def _json_value(value: Any) -> Any:
    if isinstance(value, Enum):
        return value.value
    if is_dataclass(value) and not isinstance(value, type):
        return _json_value(asdict(value))
    if isinstance(value, Mapping):
        return {str(key): _json_value(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_json_value(item) for item in value]
    return value


def _request_sync(
    method: str,
    url: str,
    headers: Mapping[str, str] | None,
    body: Any,
    timeout_seconds: float,
) -> RawResponse:
    request_headers = dict(headers or {})
    payload: bytes | None = None
    if body is not None:
        payload = json.dumps(
            _json_value(body), ensure_ascii=False, separators=(",", ":")
        ).encode("utf-8")
        request_headers["Content-Type"] = "application/json"
        request_headers["Content-Length"] = str(len(payload))
    request = Request(url, data=payload, headers=request_headers, method=method)
    try:
        with urlopen(request, timeout=timeout_seconds) as response:
            return RawResponse(
                status=int(response.status),
                body=response.read().decode("utf-8", errors="replace"),
            )
    except HTTPError as error:
        return RawResponse(
            status=int(error.code),
            body=error.read().decode("utf-8", errors="replace"),
        )


def _request_stream_sync(
    method: str,
    url: str,
    headers: Mapping[str, str] | None,
    timeout_seconds: float,
) -> StreamResponse:
    request = Request(url, headers=dict(headers or {}), method=method)
    try:
        response = urlopen(request, timeout=timeout_seconds)
        return StreamResponse(int(response.status), response.headers, response)
    except HTTPError as error:
        return StreamResponse(int(error.code), error.headers, error)


async def request(
    *,
    method: str,
    url: str,
    headers: Mapping[str, str] | None = None,
    body: Any = None,
    timeout_seconds: float = 30,
) -> RawResponse:
    return await asyncio.to_thread(
        _request_sync, method, url, headers, body, timeout_seconds
    )


async def request_stream(
    *,
    method: str,
    url: str,
    headers: Mapping[str, str] | None = None,
    timeout_seconds: float = 300,
) -> StreamResponse:
    return await asyncio.to_thread(
        _request_stream_sync, method, url, headers, timeout_seconds
    )


class HttpClient:
    """HTTP client with a default timeout and BK-CI result decoders.

    ``transport`` is an optional synchronous test adapter with the same keyword arguments as
    :func:`_request_sync`.
    """

    def __init__(
        self,
        *,
        timeout_seconds: float = 30,
        transport: RawTransport | None = None,
    ) -> None:
        self.timeout_seconds = timeout_seconds
        self._transport = transport

    async def request_raw(
        self,
        *,
        method: str,
        url: str,
        headers: Mapping[str, str] | None = None,
        body: Any = None,
    ) -> RawResponse:
        if self._transport is not None:
            return await asyncio.to_thread(
                self._transport,
                method=method,
                url=url,
                headers=headers,
                body=body,
                timeout_seconds=self.timeout_seconds,
            )
        return await request(
            method=method,
            url=url,
            headers=headers,
            body=body,
            timeout_seconds=self.timeout_seconds,
        )

    async def into_devops_result(
        self,
        *,
        method: str,
        url: str,
        headers: Mapping[str, str] | None = None,
        body: Any = None,
    ) -> DevopsResult[Any]:
        raw = await self.request_raw(
            method=method, url=url, headers=headers, body=body
        )
        parsed = _parse_result(raw, url)
        return DevopsResult(
            data=parsed.get("data"),
            status=int(parsed.get("status", -1)),
            message=str(parsed.get("message", "")),
        )

    async def into_agent_result(
        self,
        *,
        method: str,
        url: str,
        headers: Mapping[str, str] | None = None,
        body: Any = None,
    ) -> AgentResult[Any]:
        raw = await self.request_raw(
            method=method, url=url, headers=headers, body=body
        )
        parsed = _parse_result(raw, url)
        return AgentResult(
            data=parsed.get("data"),
            status=int(parsed.get("status", -1)),
            message=str(parsed.get("message", "")),
            agent_status=str(parsed.get("agentStatus", "")),
        )


def _parse_result(raw: RawResponse, url: str) -> dict[str, Any]:
    try:
        value = json.loads(raw.body)
    except (TypeError, json.JSONDecodeError) as error:
        raise ValueError(
            f"parse result error, url={url} status={raw.status} body={raw.body[:500]}"
        ) from error
    if not isinstance(value, dict):
        raise ValueError(
            f"parse result error, url={url} status={raw.status} body={raw.body[:500]}"
        )
    return value

